api/agent/drivers/docker/imageCache.go

```diff
@@ -0,0 +1,210 @@
+/*
+ImageCache holds all the logic for calculating what docker images can be removed from the running agent.
+The last used time and the number of uses are both taken into account to calculate a score (timeSinceLastUse/uses)
+The higher the score the more evicitable the image is.
+
+ImageCache also provides a method to "lock" an image, insuring it is never deleted. To do so a Lock is called with
+the image ID to lock, as well as a token. The token is then added to a set of tokens attached to that entry.
+The set is a map of *interface -> *interface where both values are the same.
+*/
+
+package docker
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	d "github.com/fsouza/go-dockerclient"
+	"github.com/sirupsen/logrus"
+)
+
+// Cache is a list of Entries. Stored as type EntryByAge so that sort works.
+// The mutex is present to control access access the list of images and their locks
+// The Cache may be used in concurrent contexts
+type Cache struct {
+	mu      sync.Mutex
+	cache   EntryByAge
+	maxSize int64
+}
+
+// Entry contains when the image was last used, how many times the image has been used,
+// The image metadata, and a set of tokens that are locking the image in question.
+type Entry struct {
+	lastUsed time.Time
+	locked   map[*interface{}]*interface{}
+	uses     int64
+	image    d.APIImages
+}
+
+// The score is the time since last use divided by the number of total uses.
```

Comment:
comments on exported methods should start with the name of the method (and matching casing), eg `Score is the time since...` to make lint happy (I appreciate the comments here a lot)